### PR TITLE
feat: add support for secant functionality `math/base/special/sec`

### LINKED ISSUE #225: [RFC]: add support for computing the secant (sec)
## Checklist

> Please ensure the following tasks are completed before submitting a feature request.

-   [x] Read and understood the [Code of Conduct][code-of-conduct].
-   [x] Searched for existing issues and pull requests.
-   [x] The issue name begins with `RFC:`.

## Description

> Description of the feature request.

This RFC proposes adding support for computing the secant. The initial implementation is straightforward:

```javascript
return 1.0 / cos( x );
```

Package: `@stdlib/math/base/special/sec`
Alias: `base.sec`

## Related Issues

> Does this feature request have any related issues?

No.

## Questions

> Any questions for reviewers?

No.

## Other

> Any other information relevant to this feature request? This may include screenshots, references, sample output, and/or implementation notes.

Reference implementations:

-   [Julia](https://github.com/JuliaLang/julia/blob/dee7d77c199c4776cf4e06afb7ce920f09ad3c8c/base/special/trig.jl#L958)

[code-of-conduct]: https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md

### PR DESCRIPTION
**Pull request by AbhijitRaut04**
Resolves #225.

## Description

> What is the purpose of this pull request?

This pull request:

-   Calculates the secant for an input in radians
- Also has the benchmarks, tests and examples

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #225 
- fixes #225 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.
Screenshot 2024-10-21 192121
Screenshot 2024-10-21 192140
Screenshot 2024-10-21 204945

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/math/base/special/cos/test/fixtures/julia/huge_negative.json
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/cos/test/fixtures/julia/huge_negative.json:1]: Why was this file changed?

File: lib/node_modules/@stdlib/math/base/special/sec/lib/index.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+* v = sec( 3.141592653589793/3.0 );
+* // returns ~2.0
+*
+* v = sec( NaN );
+* // returns NaN
+*/
+
+// MODULES //
+
+var main = require( '../main.js' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sec/lib/index.js]:
```suggestion
var main = require( './main.js' );
```

File: lib/node_modules/@stdlib/math/base/special/sec/test/test.js [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+var veryLargePositive = require( '@stdlib/math/base/special/sec/test/fixtures/julia/very_large_positive.json' );
+var largeNegative = require( '@stdlib/math/base/special/sec/test/fixtures/julia/large_negative.json' );
+var largePositive = require( '@stdlib/math/base/special/sec/test/fixtures/julia/large_positive.json' );
+var mediumNegative = require( '@stdlib/math/base/special/sec/test/fixtures/julia/medium_negative.json' );
+var mediumPositive = require( '@stdlib/math/base/special/sec/test/fixtures/julia/medium_positive.json' );
+var smallNegative = require( '@stdlib/math/base/special/sec/test/fixtures/julia/small_negative.json' );
+var smallPositive = require( '@stdlib/math/base/special/sec/test/fixtures/julia/small_positive.json' );
+var smaller = require( '@stdlib/math/base/special/sec/test/fixtures/julia/smaller.json' );
+var tinyNegative = require( '@stdlib/math/base/special/sec/test/fixtures/julia/tiny_negative.json' );
+var tinyPositive = require( '@stdlib/math/base/special/sec/test/fixtures/julia/tiny_positive.json' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sec/test/test.js]: Please see other packages. These should all be relative paths.

File: lib/node_modules/@stdlib/math/base/special/sec/test/test.js [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+	var x;
+	var y;
+	var i;
+
+	x = hugeNegative.x;
+	expected = hugeNegative.expected;
+
+	for (i = 0; i < x.length; i++) {
+		y = sec(x[i]);
+		console.log(y)
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sec/test/test.js]: Is this intended?

File: lib/node_modules/@stdlib/math/base/special/sec/test/test.js [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+	var x;
+	var y;
+	var i;
+
+	x = hugePositive.x;
+	expected = hugePositive.expected;
+
+	for (i = 0; i < x.length; i++) {
+		y = sec(x[i]);
+		if (y === expected[i]) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sec/test/test.js]: Please honor the project's spacing conventions here and elsewhere throughout this PR.

File: lib/node_modules/@stdlib/math/base/special/sec/src/main.c [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+#include "stdlib/math/base/special/cos.h"
+
+/**
+* Evaluates the cosecant of a number.
+*
+* @param x    input value (in radians)
+* @return     secant
+*
+* @example
+* double y = stdlib_base_cos( 3.141592653589793 / 2.0 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sec/src/main.c]:
```suggestion
* double y = stdlib_base_sec( 3.141592653589793 / 2.0 );
```

File: lib/node_modules/@stdlib/math/base/special/sec/docs/repl.txt [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+    Returns
+    -------
+    y: number
+        Secant.
+
+    Examples
+    --------
+    > var y = {{alias}}( 0.0 )
+    1
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sec/docs/repl.txt]:
```suggestion
    1.0
```

File: lib/node_modules/@stdlib/math/base/special/sec/benchmark/benchmark.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+			b.fail( 'should not return NaN' );
+		}
+	}
+	b.toc();
+	if ( isnan( y ) ) {
+		b.fail( 'should not return NaN' );
+	}
+	b.pass( 'benchmark finished' );
+	b.end();
+});
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sec/benchmark/benchmark.js:73]:
```suggestion
```

File: lib/node_modules/@stdlib/math/base/special/sec/README.md [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+## sec( x )
+
+Evaluates the [secant][trigonometric-functions] of `x` (in radians).
+
+```javascript
+var v = sec( 0.0 );
+// returns 1.0
+
+v = sec( 3.141592653589793/2.0 );
+// returns Infinity
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/sec/README.md]: Is the value actually supposed to equal infinity? Given finite precision, IIRC, `cos(pi/2)` does not equal `0`.

File: lib/node_modules/@stdlib/math/base/special/sec/docs/repl.txt [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+        Secant.
+
+    Examples
+    --------
+    > var y = {{alias}}( 0.0 )
+    1.0
+    > y = {{alias}}( {{alias:@stdlib/constants/float64/pi}}/2.0 )
+    ~Infinity
+    > y = {{alias}}( -{{alias:@stdlib/constants/float64/pi}}/3.0 )
+    ~2.0
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/sec/docs/repl.txt]: I propose matching the return annotation values used elsewhere:

```suggestion
    > y = {{alias}}( {{alias:@stdlib/constants/float64/pi}}/2.0 )
    16331239353195370
    > y = {{alias}}( -{{alias:@stdlib/constants/float64/pi}}/3.0 )
    ~1.9999999999999996
```

File: lib/node_modules/@stdlib/math/base/special/sec/README.md [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+
+<section class="related">
+
+* * *
+
+## See Also
+
+-   <span class="package-name">[`@stdlib/math/base/special/cot`][@stdlib/math/base/special/cot]</span><span class="delimiter">: </span><span class="description">evaluate the cotangent of a number.</span>
+-   <span class="package-name">[`@stdlib/math/base/special/cos`][@stdlib/math/base/special/cos]</span><span class="delimiter">: </span><span class="description">compute the cosine of a number.</span>
+
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/sec/README.md]: Let's remove this since it will be auto-generated.

```suggestion
```

File: lib/node_modules/@stdlib/math/base/special/sec/README.md [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+<section class="links">
+
+[trigonometric-functions]: https://en.wikipedia.org/wiki/Trigonometric_functions
+
+<!-- <related-links> -->
+
+[@stdlib/math/base/special/cot]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/cot
+
+[@stdlib/math/base/special/cos]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/cos
+
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/sec/README.md]:
```suggestion
```

**Review by Planeshifter — APPROVED** [review]
PR is in good shape, I think. Left some minor comments but it should be ready for merging otherwise.

**Comment by AbhijitRaut04** [comment]
Hi @Planeshifter and @kgryte,

I've made the requested changes and pushed them. Please let me know if anything else is needed. Thank you so much for your guidance throughout the process! 😊

This was my first contribution to an open-source project, and I’m thrilled to see this PR getting merged. Once again, thank you for your support!

File: lib/node_modules/@stdlib/math/base/special/sec/src/addon.c [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include "stdlib/math/base/special/sec.h"
+#include "stdlib/math/base/napi/unary.h"
+
+STDLIB_MATH_BASE_NAPI_MODULE_D_D( stdlib_base_sec )
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/sec/src/addon.c:23]:
```suggestion
// cppcheck-suppress shadowFunction
STDLIB_MATH_BASE_NAPI_MODULE_D_D( stdlib_base_sec )
```

File: lib/node_modules/@stdlib/math/base/special/sec/lib/main.js [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+* @param {number} x - input value (in radians)
+* @returns {number} secant
+*
+* @example
+* var v = sec( 0.0 );
+* // returns 1.0
+*
+* @example
+* var v = sec( 3.141592653589793/2.0 );
+* // returns 16331239353195370
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/sec/lib/main.js]:
```suggestion
* // returns 16331239353195370.0
```

File: lib/node_modules/@stdlib/math/base/special/sec/lib/index.js [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+* @module @stdlib/math/base/special/sec
+*
+* @example
+* var sec = require( '@stdlib/math/base/special/sec' );
+*
+* var v = sec( 0.0 );
+* // returns 1.0
+*
+* v = sec( 3.141592653589793/2.0 );
+* // returns 16331239353195370
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/sec/lib/index.js]:
```suggestion
* // returns 16331239353195370.0
```

File: lib/node_modules/@stdlib/math/base/special/sec/docs/types/index.d.ts [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+* @param x - input value (in radians)
+* @returns secant
+*
+* @example
+* var v = sec( 0.0 );
+* // returns 1.0
+*
+* @example
+* var v = sec( 3.141592653589793/2.0 );
+* // returns 16331239353195370
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/sec/docs/types/index.d.ts]:
```suggestion
* // returns 16331239353195370.0
```

File: lib/node_modules/@stdlib/math/base/special/sec/README.md [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+## sec( x )
+
+Evaluates the [secant][trigonometric-functions] of `x` (in radians).
+
+```javascript
+var v = sec( 0.0 );
+// returns 1.0
+
+v = sec( 3.141592653589793/2.0 );
+// returns 16331239353195370
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/sec/README.md]:
```suggestion
// returns 16331239353195370.0
```

File: lib/node_modules/@stdlib/math/base/special/sec/docs/repl.txt [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+    -------
+    y: number
+        Secant.
+
+    Examples
+    --------
+    > var y = {{alias}}( 0.0 )
+    1.0
+    > y = {{alias}}( {{alias:@stdlib/constants/float64/pi}}/2.0 )
+    ~16331239353195370
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/sec/docs/repl.txt]:
```suggestion
    16331239353195370.0
```

File: lib/node_modules/@stdlib/math/base/special/sec/lib/native.js [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+* @param {number} x - input value (in radians)
+* @returns {number} secant
+*
+* @example
+* var v = sec( 0.0 );
+* // returns 1.0
+*
+* @example
+* var v = sec( 3.141592653589793 / 2.0 );
+* // returns 16331239353195370
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/sec/lib/native.js]:
```suggestion
* // returns 16331239353195370.0
```

**Review by Planeshifter — APPROVED** [review]

File: lib/node_modules/@stdlib/math/base/special/sec/src/main.c [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+* Evaluates the cosecant of a number.
+*
+* @param x    input value (in radians)
+* @return     secant
+*
+* @example
+* double y = stdlib_base_sec( 3.141592653589793 / 2.0 );
+* // returns 16331239353195370
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/sec/src/main.c]:
```suggestion
* // returns 16331239353195370.0
```

File: lib/node_modules/@stdlib/math/base/special/sec/test/test.js [hunk 20]
```diff
@@ -0,0 +1,396 @@
...
+// MODULES //
+
+var tape = require( 'tape' );
+var isnan = require( '@stdlib/math/base/assert/is-nan' );
+var abs = require( '@stdlib/math/base/special/abs' );
+var PI = require( '@stdlib/constants/float64/pi' );
+var EPS = require( '@stdlib/constants/float64/eps' );
+var PINF = require( '@stdlib/constants/float64/pinf' );
+var NINF = require( '@stdlib/constants/float64/ninf' );
+var sec = require( '../lib' );
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/sec/test/test.js]:
```suggestion
var sec = require( './../lib' );
```

**Comment by AbhijitRaut04** [comment]
@Planeshifter Should I need to push the above changes ?

**Comment by Planeshifter** [comment]
@AbhijitRaut04 All good, I was free to apply them and merged your PR! :tada: 

Thanks again for your efforts.
